File: ts/ou_bench.ts
/*
  Unified OU benchmark (Bun / TypeScript)

  Algorithms intentionally match the Rust/Zig/C versions in this repo:
  - PRNG: xorshift128 (u32) seeded with splitmix32
  - Uniform: 53-bit double from two u32 draws
  - Normal: Marsaglia polar method with cached spare
  - OU: Euler update with precomputed a,b and diffusion coefficient

  Run:
    bun run ou_bench.ts --n=500000 --runs=1000 --warmup=5 --seed=1
*/

type Args = {
  n: number;
  runs: number;
  warmup: number;
  seed: number;
};

function parseArgs(argv: string[]): Args {
  // Defaults match the blog's parameters.
  const args: Args = { n: 500_000, runs: 1000, warmup: 5, seed: 1 };

  for (const raw of argv) {
    if (!raw.startsWith("--")) continue;
    const eq = raw.indexOf("=");
    const key = eq >= 0 ? raw.slice(2, eq) : raw.slice(2);
    const val = eq >= 0 ? raw.slice(eq + 1) : "";
    const num = val.length ? Number(val) : NaN;

    switch (key) {
      case "n":
        if (!Number.isFinite(num) || num < 2) throw new Error("--n must be >= 2");
        args.n = Math.floor(num);
        break;
      case "runs":
        if (!Number.isFinite(num) || num < 1) throw new Error("--runs must be >= 1");
        args.runs = Math.floor(num);
        break;
      case "warmup":
        if (!Number.isFinite(num) || num < 0) throw new Error("--warmup must be >= 0");
        args.warmup = Math.floor(num);
        break;
      case "seed":
        if (!Number.isFinite(num) || num < 0) throw new Error("--seed must be >= 0");
        args.seed = Math.floor(num) >>> 0; // keep in u32 range
        break;
      default:
        // ignore unknown args
        break;
    }
  }
  return args;
}

// ---- PRNG: splitmix32 seeding + xorshift128 ----

function splitmix32_next(state: { s: number }): number {
  // All arithmetic is 32-bit.
  state.s = (state.s + 0x9e3779b9) | 0;
  let z = state.s | 0;
  z = Math.imul(z ^ (z >>> 16), 0x85ebca6b) | 0;
  z = Math.imul(z ^ (z >>> 13), 0xc2b2ae35) | 0;
  z = (z ^ (z >>> 16)) | 0;
  return z >>> 0;
}

class XorShift128 {
  private x: number;
  private y: number;
  private z: number;
  private w: number;

  constructor(seed: number) {
    // seed via splitmix32 into 4 non-zero-ish states
    const st = { s: seed | 0 };
    this.x = splitmix32_next(st) | 0;
    this.y = splitmix32_next(st) | 0;
    this.z = splitmix32_next(st) | 0;
    this.w = splitmix32_next(st) | 0;

    // Avoid the all-zero state (extremely unlikely, but possible).
    if ((this.x | this.y | this.z | this.w) === 0) {
      this.w = 1;
    }
  }

  nextU32(): number {
    // Marsaglia xorshift128 (returns u32)
    const t = (this.x ^ (this.x << 11)) | 0;
    this.x = this.y;
    this.y = this.z;
    this.z = this.w;
    this.w = (this.w ^ (this.w >>> 19) ^ t ^ (t >>> 8)) | 0;
    return this.w >>> 0;
  }

  nextF64(): number {
    // 53-bit uniform in [0,1) from two u32 draws (exact in JS number range)
    const a = this.nextU32();
    const b = this.nextU32();
    const u = (a >>> 5) * 67108864 + (b >>> 6); // (a>>5)<<26 + (b>>6)
    return u * (1.0 / 9007199254740992.0); // 2^53
  }
}

// ---- Normal: Marsaglia polar with cached spare ----

class NormalPolar {
  private hasSpare = false;
  private spare = 0.0;

  nextStandard(rng: XorShift128): number {
    if (this.hasSpare) {
      this.hasSpare = false;
      return this.spare;
    }
    while (true) {
      const u = 2.0 * rng.nextF64() - 1.0;
      const v = 2.0 * rng.nextF64() - 1.0;
      const s = u * u + v * v;
      if (s > 0.0 && s < 1.0) {
        const m = Math.sqrt((-2.0 * Math.log(s)) / s);
        this.spare = v * m;
        this.hasSpare = true;
        return u * m;
      }
    }
  }
}

// ---- Benchmark core ----

function nowMs(): number {
  // Bun supports performance.now()
  return performance.now();
}

function main(): void {
  const { n, runs, warmup, seed } = parseArgs(process.argv.slice(2));

  const T = 1.0;
  const theta = 1.0;
  const mu = 0.0;
  const sigma = 0.1;

  const dt = T / n;
  const a = 1.0 - theta * dt;
  const b = theta * mu * dt;
  const diff = sigma * Math.sqrt(dt);

  const gn = new Float64Array(n - 1);
  const ou = new Float64Array(n);

  // Warmup (JIT + caches)
  {
    const rng = new XorShift128(seed);
    const norm = new NormalPolar();
    for (let r = 0; r < warmup; r++) {
      // generate gn
      for (let i = 0; i < n - 1; i++) {
        gn[i] = diff * norm.nextStandard(rng);
      }
      // simulate OU
      let x = 0.0;
      ou[0] = x;
      for (let i = 1; i < n; i++) {
        x = a * x + b + gn[i - 1];
        ou[i] = x;
      }
      // checksum readback (forces stores)
      let s = 0.0;
      for (let i = 0; i < n; i++) s += ou[i];
      // prevent whole warmup from being dead
      if (s === 123456789.0) console.log("impossible");
    }
  }

  // Force GC before timed runs to reduce variance
  if (typeof Bun !== "undefined" && Bun.gc) {
    Bun.gc(true);
  }

  // Timed runs
  const rng = new XorShift128(seed);
  const norm = new NormalPolar();

  let totalMs = 0.0;
  let totalGenMs = 0.0;
  let totalSimMs = 0.0;
  let totalChkMs = 0.0;

  let minMs = Number.POSITIVE_INFINITY;
  let maxMs = 0.0;
  const runTimes: number[] = [];

  let checksum = 0.0;

  for (let r = 0; r < runs; r++) {
    const t0 = nowMs();

    // (1) generate Gaussian increments
    for (let i = 0; i < n - 1; i++) {
      gn[i] = diff * norm.nextStandard(rng);
    }
    const t1 = nowMs();

    // (2) simulate OU
    let x = 0.0;
    ou[0] = x;
    for (let i = 1; i < n; i++) {
      x = a * x + b + gn[i - 1];
      ou[i] = x;
    }
    const t2 = nowMs();

    // (3) checksum: full readback to prevent dead-store elimination
    let s = 0.0;
    for (let i = 0; i < n; i++) s += ou[i];
    checksum += s;
    const t3 = nowMs();

    const genMs = t1 - t0;
    const simMs = t2 - t1;
    const chkMs = t3 - t2;
    const runMs = t3 - t0;

    totalGenMs += genMs;
    totalSimMs += simMs;
    totalChkMs += chkMs;
    totalMs += runMs;
    runTimes.push(runMs);

    if (runMs < minMs) minMs = runMs;
    if (runMs > maxMs) maxMs = runMs;
  }

  const avgMs = totalMs / runs;
  runTimes.sort((a, b) => a - b);
  const medianMs = runs % 2 === 1
    ? runTimes[Math.floor(runs / 2)]
    : (runTimes[runs / 2 - 1] + runTimes[runs / 2]) / 2;

  console.log("== OU benchmark (TypeScript/Bun, unified algorithms) ==");
  console.log(`n=${n} runs=${runs} warmup=${warmup} seed=${seed}`);
  console.log(`total_s=${(totalMs / 1000).toFixed(6)}`);
  console.log(`avg_ms=${avgMs.toFixed(6)} median_ms=${medianMs.toFixed(6)} min_ms=${minMs.toFixed(6)} max_ms=${maxMs.toFixed(6)}`);
  console.log(
    `breakdown_s gen_normals=${(totalGenMs / 1000).toFixed(6)} simulate=${(totalSimMs / 1000).toFixed(6)} checksum=${(totalChkMs / 1000).toFixed(6)}`
  );
  console.log(`checksum=${checksum.toPrecision(17)}`);
}

main();

File: c/ou_bench.c
/*
  Unified OU benchmark (C)

  Algorithms intentionally match TS/Rust/Zig in this repo:
  - PRNG: xorshift128 (u32) seeded via splitmix32
  - Uniform: 53-bit double from two u32 draws
  - Normal: Marsaglia polar method with cached spare
  - OU: Euler update with precomputed a,b and diffusion coefficient

  Build:
    cc -O3 -march=native -std=c11 ou_bench.c -lm -o ou_bench_c

  Run:
    ./ou_bench_c --n=500000 --runs=1000 --warmup=5 --seed=1
*/

#define _POSIX_C_SOURCE 199309L
#include <stdint.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <math.h>

typedef struct {
    uint32_t s;
} splitmix32_t;

static inline uint32_t splitmix32_next(splitmix32_t *st) {
    st->s += 0x9E3779B9u;
    uint32_t z = st->s;
    z = (z ^ (z >> 16)) * 0x85EBCA6Bu;
    z = (z ^ (z >> 13)) * 0xC2B2AE35u;
    z = z ^ (z >> 16);
    return z;
}

typedef struct {
    uint32_t x, y, z, w;
} xorshift128_t;

static inline xorshift128_t xorshift128_new(uint32_t seed) {
    splitmix32_t sm = { seed };
    xorshift128_t rng;
    rng.x = splitmix32_next(&sm);
    rng.y = splitmix32_next(&sm);
    rng.z = splitmix32_next(&sm);
    rng.w = splitmix32_next(&sm);
    if ((rng.x | rng.y | rng.z | rng.w) == 0u) {
        rng.w = 1u;
    }
    return rng;
}

static inline uint32_t xorshift128_next_u32(xorshift128_t *rng) {
    // Marsaglia xorshift128 (32-bit)
    uint32_t t = rng->x ^ (rng->x << 11);
    rng->x = rng->y;
    rng->y = rng->z;
    rng->z = rng->w;
    rng->w = rng->w ^ (rng->w >> 19) ^ t ^ (t >> 8);
    return rng->w;
}

static inline double xorshift128_next_f64(xorshift128_t *rng) {
    // 53-bit uniform in [0,1) from two u32 draws.
    uint32_t a = xorshift128_next_u32(rng);
    uint32_t b = xorshift128_next_u32(rng);
    uint64_t u = ((uint64_t)(a >> 5) << 26) | (uint64_t)(b >> 6);
    return (double)u * (1.0 / 9007199254740992.0); // 2^53
}

typedef struct {
    int has_spare;
    double spare;
} normal_polar_t;

static inline void normal_polar_init(normal_polar_t *n) {
    n->has_spare = 0;
    n->spare = 0.0;
}

static inline double normal_polar_next(normal_polar_t *n, xorshift128_t *rng) {
    if (n->has_spare) {
        n->has_spare = 0;
        return n->spare;
    }
    for (;;) {
        double u = 2.0 * xorshift128_next_f64(rng) - 1.0;
        double v = 2.0 * xorshift128_next_f64(rng) - 1.0;
        double s = u*u + v*v;
        if (s > 0.0 && s < 1.0) {
            double m = sqrt((-2.0 * log(s)) / s);
            n->spare = v * m;
            n->has_spare = 1;
            return u * m;
        }
    }
}

static inline uint64_t now_ns(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (uint64_t)ts.tv_sec * 1000000000ull + (uint64_t)ts.tv_nsec;
}

typedef struct {
    size_t n;
    size_t runs;
    size_t warmup;
    uint32_t seed;
} args_t;

static args_t parse_args(int argc, char **argv) {
    args_t a;
    a.n = 500000;
    a.runs = 1000;
    a.warmup = 5;
    a.seed = 1;

    for (int i = 1; i < argc; i++) {
        const char *s = argv[i];
        if (strncmp(s, "--n=", 4) == 0) {
            long long v = atoll(s + 4);
            if (v < 2) { fprintf(stderr, "--n must be >= 2\n"); exit(1); }
            a.n = (size_t)v;
        } else if (strncmp(s, "--runs=", 7) == 0) {
            long long v = atoll(s + 7);
            if (v < 1) { fprintf(stderr, "--runs must be >= 1\n"); exit(1); }
            a.runs = (size_t)v;
        } else if (strncmp(s, "--warmup=", 9) == 0) {
            long long v = atoll(s + 9);
            if (v < 0) { fprintf(stderr, "--warmup must be >= 0\n"); exit(1); }
            a.warmup = (size_t)v;
        } else if (strncmp(s, "--seed=", 7) == 0) {
            unsigned long long v = strtoull(s + 7, NULL, 10);
            a.seed = (uint32_t)(v & 0xFFFFFFFFu);
        }
    }

    return a;
}

int main(int argc, char **argv) {
    args_t args = parse_args(argc, argv);

    const double T = 1.0;
    const double theta = 1.0;
    const double mu = 0.0;
    const double sigma = 0.1;

    const size_t n = args.n;

    const double dt = T / (double)n;
    const double a = 1.0 - theta * dt;
    const double b = theta * mu * dt;
    const double diff = sigma * sqrt(dt);

    double *gn = (double*)malloc((n - 1) * sizeof(double));
    double *ou = (double*)malloc(n * sizeof(double));
    if (!gn || !ou) {
        fprintf(stderr, "allocation failed\n");
        return 1;
    }

    // Warmup
    {
        xorshift128_t rng = xorshift128_new(args.seed);
        normal_polar_t norm;
        normal_polar_init(&norm);

        for (size_t r = 0; r < args.warmup; r++) {
            for (size_t i = 0; i < n - 1; i++) {
                gn[i] = diff * normal_polar_next(&norm, &rng);
            }
            double x = 0.0;
            ou[0] = x;
            for (size_t i = 1; i < n; i++) {
                x = a * x + b + gn[i - 1];
                ou[i] = x;
            }
            double s = 0.0;
            for (size_t i = 0; i < n; i++) s += ou[i];
            if (s == 123456789.0) printf("impossible\n");
        }
    }

    // Timed runs
    xorshift128_t rng = xorshift128_new(args.seed);
    normal_polar_t norm;
    normal_polar_init(&norm);

    double total_s = 0.0;
    double total_gen_s = 0.0;
    double total_sim_s = 0.0;
    double total_chk_s = 0.0;

    double min_s = 1e300;
    double max_s = 0.0;
    double *run_times = (double*)malloc(args.runs * sizeof(double));
    if (!run_times) {
        fprintf(stderr, "allocation failed\n");
        return 1;
    }

    double checksum = 0.0;

    for (size_t r = 0; r < args.runs; r++) {
        uint64_t t0 = now_ns();

        for (size_t i = 0; i < n - 1; i++) {
            gn[i] = diff * normal_polar_next(&norm, &rng);
        }
        uint64_t t1 = now_ns();

        double x = 0.0;
        ou[0] = x;
        for (size_t i = 1; i < n; i++) {
            x = a * x + b + gn[i - 1];
            ou[i] = x;
        }
        uint64_t t2 = now_ns();

        double s = 0.0;
        for (size_t i = 0; i < n; i++) s += ou[i];
        checksum += s;
        uint64_t t3 = now_ns();

        double gen = (double)(t1 - t0) * 1e-9;
        double sim = (double)(t2 - t1) * 1e-9;
        double chk = (double)(t3 - t2) * 1e-9;
        double run = (double)(t3 - t0) * 1e-9;

        total_gen_s += gen;
        total_sim_s += sim;
        total_chk_s += chk;
        total_s += run;
        run_times[r] = run;

        if (run < min_s) min_s = run;
        if (run > max_s) max_s = run;
    }

    // Sort run_times for median
    for (size_t i = 0; i < args.runs - 1; i++) {
        for (size_t j = i + 1; j < args.runs; j++) {
            if (run_times[j] < run_times[i]) {
                double tmp = run_times[i];
                run_times[i] = run_times[j];
                run_times[j] = tmp;
            }
        }
    }
    double median_s = (args.runs % 2 == 1)
        ? run_times[args.runs / 2]
        : (run_times[args.runs / 2 - 1] + run_times[args.runs / 2]) / 2.0;

    double avg_ms = (total_s / (double)args.runs) * 1000.0;
    double median_ms = median_s * 1000.0;
    double min_ms = min_s * 1000.0;
    double max_ms = max_s * 1000.0;

    printf("== OU benchmark (C, unified algorithms) ==\n");
    printf("n=%zu runs=%zu warmup=%zu seed=%u\n", args.n, args.runs, args.warmup, args.seed);
    printf("total_s=%.6f\n", total_s);
    printf("avg_ms=%.6f median_ms=%.6f min_ms=%.6f max_ms=%.6f\n", avg_ms, median_ms, min_ms, max_ms);
    printf("breakdown_s gen_normals=%.6f simulate=%.6f checksum=%.6f\n", total_gen_s, total_sim_s, total_chk_s);
    printf("checksum=%.17g\n", checksum);

    free(gn);
    free(ou);
    free(run_times);
    return 0;
}
